tests/python_package_test/test_dask.py

```diff
@@ -1255,3 +1255,59 @@ def test_sklearn_integration(estimator, check, client):
 def test_parameters_default_constructible(estimator):
     name, Estimator = estimator.__class__.__name__, estimator.__class__
     sklearn_checks.check_parameters_default_constructible(name, Estimator)
+
+
+@pytest.mark.parametrize('task', ['binary_classification', 'multi-class_classification', 'regression'])
```

Comment:
can you please adding ranking tests as well?

Reply:
actually this change also makes me realize that the other tests using `@pytest.mark.parametrize('task', tasks)` are not currently testing multi-class classification.

Are you interested in submitting a separate PR to add `multi-class-classification` to `tasks`? https://github.com/microsoft/LightGBM/blob/6356e659af74909aa77eb87d54b79f0373772509/tests/python_package_test/test_dask.py#L41

Doing that would improve our test coverage AND reduce the risk of mistakes like this where a task if forgotten.

Reply:
Sure. Should this modify the `_create_data` function to add an objective of `multi-class-classification`?

Reply:
yes please! based on this discussion: https://github.com/microsoft/LightGBM/pull/4024#discussion_r583795293

when you do that, I think it would make sense to remove the `centers` argument from that function, and just use the `objective` to set `centers=2` or `centers=3`. We don't have any tests right now that use more than 3 classes or that really care about the exact numeric values of the cluster centers coming out of `make_blobs()`.

Reply:
Hi James. I'm having trouble with the tests for ranking. The check for the expected mean is because if you have 25 samples in one leaf with value 0 and 75 samples in another leaf with value 1 you expect your mean to be (25 * 0 + 75 * 1) / (25 + 75), right? Which I guess would be roughly equal to checking that there are 25 predictions with the value 0 and 75 with the value 1. But for ranking I'm seeing that these don't match:
![image](https://user-images.githubusercontent.com/8473587/110878101-30cc3680-82a0-11eb-8756-24aa95209c48.png)

Reply:
btw this doesn't happen with local ranker.

Reply:
I don't understand what you mean by "don't match" and "having trouble", sorry. Where are you getting the numbers 75 and 25 from?

Reply:
By having trouble I mean the tests aren't passing because the expected mean isn't the actual mean. The 75 and 25 in the image are the actual counts for the prediction values, so we expect to have 77 samples with prediction -0.025374 and 23 samples with prediction 0.072166 but we get 75 and 25 respectively.

Reply:
oh wow! Ok yeah that was not clear to me, thanks for rephrasing it. That definitely seems problematic.

However, I want to clarify something...you're using `weight`, not `count`, to get the expected mean, right?

Reply:
Yes, I used the counts here for clarity but the normalized weights are roughly the same (0.772, 0.228)

Reply:
Ok, got it! This definitely seems like a bug. Does it happen on every test for the ranker, and only for the ranker?

If so, can you open a separate issue with a minimal reproducible example? You could use https://github.com/microsoft/LightGBM/blob/master/examples/python-guide/dask/ranking.py if you need a simpler data setup for ranking.

This seems like a bug and I have a few ideas of where it could be but I don't expect that it's specific to Dask or to the changes in this PR, so I think it would be better to discuss it in a separate issue.

Reply:
Every test for the ranker fails, but I just realized that I wasn't adding sample weight to the other tasks and if I do then they all fail. This isn't right, is it?

Reply:
Is the reverse true? If you do unweighted training for all tasks, do the tests pass?

If so, then I think these tests should use unweighted training, and that the "counts of unique preds disagrees with the results of `trees_to_dataframe()`" thing is a separate issue.

Reply:
Yes, without using weights the tasks succeed except for the ranker.

Reply:
oh that's different than what I understood before. I thought your last comment meant "I noticed that ranker was the only task that I was doing weighted training for".

Ok, I think the challenge here is that the `weight` column in the tree node metadata and the sample weights are related but not identical concepts. For this test, let's simplify by not using weighted training.

For the ranking test, can you just use a tolerance like `assert abs(expected_mean - preds_mean) < threshold`? And then open the separate issue from https://github.com/microsoft/LightGBM/pull/4024#discussion_r592870901 documenting the inconsistency you saw in the results and small reproducible example if possible. Based on what you've said so far I suspect you've uncovered a bug in distributed LightGBM, and I don't think that bug needs to block this PR.

Reply:
Should I compare the normalized weights from the dataframe vs the normalized counts from the unique predictions in that issue?

Reply:
Or I could compare directly the count column vs the unique counts of the predictions. It's not very clear to me why we don't use that here, why are we using the weights?

Reply:
you know what now that I think about it...it would be enough to compare the unique values of the leaf nodes to the unique values of the preds. If those are the same, you know you got the raw scores

Reply:
So the tests would check `set(raw_predictions) == set(trees_df['value'])`?

Reply:
yes

Reply:
done. please let me know what you think